test/libcommonserver/io/testcheckdirectoryiterator.cpp

```diff
@@ -0,0 +1,237 @@
+    testCheckDirectoryIteratorPermission();
+    testCheckDirectoryIteratorUnexpectedDelete();
+}
+
+void TestIo::testCheckDirectoryIteratorNonExistingPath() {
+    // Check that the directory iterator returns an error when the path does not exist
+    {
+        IoError error;
+        const DirectoryIterator it("C:\\nonexistingpath", false, error);
+
+        CPPUNIT_ASSERT_EQUAL(IoError::IoErrorNoSuchFileOrDirectory, error);
+    }
+
+    // Check that the directory iterator returns an error when the path syntax is invalid
+    {
+        IoError error;
+        const DirectoryIterator it("C:\\nonexistingpath\\*\\", false, error);
+
+        CPPUNIT_ASSERT_EQUAL(IoError::IoErrorNoSuchFileOrDirectory, error);
+    }
+}
+
+
+void TestIo::testCheckDirectoryIteratorExistingPath() {
+    // Check that the directory iterator is valid when the path is an empty directory and return EOF on first call
+    {
+        const TemporaryDirectory temporaryDirectory;
+        const SyncPath path = temporaryDirectory.path / "chekDirIt/empty_dir";
+        std::filesystem::create_directories(path);
+
+        IoError error;
+
+        DirectoryIterator it(path, false, error);
+        CPPUNIT_ASSERT_EQUAL(IoError::IoErrorSuccess, error);
+
+        DirectoryEntry entry;
+        CPPUNIT_ASSERT_EQUAL(false, it.next(entry, error));
+        CPPUNIT_ASSERT_EQUAL(IoError::IoErrorEndOfDirectory, error);
+    }
+
+    // Check that the directory iterator is valid when the path is a directory with one file and return the file on first call
+    {
+        const SyncPath directoryWithOneFile = _localTestDirPath / "test_dir_iterator/oneFile_dir";
+
+        IoError error;
+        DirectoryIterator it(directoryWithOneFile, false, error);
+        CPPUNIT_ASSERT_EQUAL(IoError::IoErrorSuccess, error);
+
+        DirectoryEntry entry;
+        CPPUNIT_ASSERT_EQUAL(true, it.next(entry, error));
+        CPPUNIT_ASSERT_EQUAL(IoError::IoErrorSuccess, error);
+
+
+        CPPUNIT_ASSERT_EQUAL(directoryWithOneFile / "testFile1.txt", entry.path());
+
+        CPPUNIT_ASSERT_EQUAL(false, it.next(entry, error));
+        CPPUNIT_ASSERT_EQUAL(IoError::IoErrorEndOfDirectory, error);
+    }
+
+    // Check that the directory iterator is valid when the path is a directory with one child directory
+    {
+        const SyncPath directoryWithOneChildDirectory = _localTestDirPath / "test_dir_iterator/oneDir_dir";
+        IoError error;
+        DirectoryIterator it(directoryWithOneChildDirectory, false, error);
+        CPPUNIT_ASSERT_EQUAL(IoError::IoErrorSuccess, error);
+
+        DirectoryEntry entry;
+        CPPUNIT_ASSERT_EQUAL(true, it.next(entry, error));
+        CPPUNIT_ASSERT_EQUAL(IoError::IoErrorSuccess, error);
+
+        CPPUNIT_ASSERT_EQUAL(directoryWithOneChildDirectory / "testDir1", entry.path());
+
+        CPPUNIT_ASSERT_EQUAL(false, it.next(entry, error));
+        CPPUNIT_ASSERT_EQUAL(IoError::IoErrorEndOfDirectory, error);
+    }
+
+
```

Comment:
You may remove this line.

Reply:
It seems to me that it can be useful to verify that the IoError thrown is the correct one, at least once during the test. I removed this assert, as it is already checked in a similar situation in the previous test. (commit: 2be9b53).